sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```
@@ -133,19 +133,23 @@ object Cast extends QueryErrorsBase {
     // to convert data of these types to Variant Objects.
     case (_, VariantType) => variant.VariantGet.checkDataType(from, allowStructsAndMaps = false)
 
+    // non-null variants can generate nulls even in ANSI mode
     case (ArrayType(fromType, fn), ArrayType(toType, tn)) =>
-      canAnsiCast(fromType, toType) && resolvableNullability(fn, tn)
+      canAnsiCast(fromType, toType) && resolvableNullability(fn || (fromType == VariantType), tn)
```

I think what we want to do is to follow `Cast.canCast`
```
    case (ArrayType(fromType, fn), ArrayType(toType, tn)) =>
      canCast(fromType, toType) &&
        resolvableNullability(fn || forceNullable(fromType, toType), tn)
```

`fromType == VariantType` makes sense, but shall we consider the deeply nested cases like array of array of variant?

---

I don't know if non-null `array<variant>` can cast to null. Yeah, I think resolveNullability is also complete in this case but only the variant case would apply in resolveNullability because every other type that fails in resolveNullability would also fail in canAnsiCast. I'll change it to resolveNullability.

---

Oh forceNullable causes it to fail for some reason. Looking into it.

---

We cannot do forceNullable because `forceNullable(String, Boolean)` returns true and String to Boolean cast is allowed in ANSI - just that if it fails, the cast throws an error instead of giving nulls.
Variant is the only data type which can legally cast from non-null to null in ANSI.